src/build.d

```diff
@@ -1480,6 +1486,75 @@ class Dependency
         }
     }
 
+     // Metadata required for parallel execution
+    private
+    {
+        shared uint builtDeps; /// amount of dependencies already built
+        bool visited; /// whether this dependency was already visited
+        Dependency[] requiredBy; /// dependencies relying on this one
+    }
+
+    /**
+    Builds all supplied targets using the global taskPool to execute buildRecursive
+    for every leaf dependency.
+
+    Params:
+        targets = dependencies to build
+    **/
+    static void build(Dependency[] targets)
+    {
+        import core.atomic: atomicLoad, atomicOp, atomicStore;
+
+        //// Recursively appends all dependencies without further dependencies to roots
+        static void findRoots(Dependency cur, ref Dependency[] roots)
```

Comment:
I think instead of using this function and the `visited` field above, you can just create a map, and check whether or not each dependency is already in the map (instead of using `visited`).  This decouples the Dependency objects from any potential sets they may be apart of.

```D
static void addToClosure(Dependency d, bool[Dependency] closure)
{
    if (d in closure) return;
    closure[d] = true;
    foreach (dep; d.deps)
    {
        dep.requiredBy ~= d;
        addToClosure(dep, closure);
    }
}
```

Reply:
I borrowed the term `Closure` from the Nix Package manager.  A closure being the "complete set" of components to support any one component, which means you not only need to deploy the component itself, but also its dependencies.

Reply:
Actually, I'm not sure the way you've written this will actually work correctly.  Suppose you have an object A that depends on B and C.  In this case, both B and C would be roots, so when either of them finish, you would immediately move on to building A correct?  It wouldn't wait for both would it?

Reply:
> Actually, I'm not sure the way you've written this will actually work correctly. Suppose you have an object A that depends on B and C. In this case, both B and C would be roots, so when either of them finish, you would immediately move on to building A correct? It wouldn't wait for both would it?

https://github.com/dlang/dmd/blob/3f5f686fa9044ca1b99e2797ed380b4162370f8b/src/build.d#L1540-L1541

The first dependency to finish increments `builtDeps` to 1 and exits because it is not equal to `A.deps.length` -> 2. The second dependency increments it to 2 and hence builds `A`.

Reply:
Ah that's right, I even saw that the other day, I just forgot.